include/rmm/mr/host/new_delete_resource.hpp

```diff
@@ -0,0 +1,109 @@
+   * and to `alignof(std::max_align_t)` otherwise.
+   *
+   * @throws std::bad_alloc When the requested `bytes` and `alignment` cannot be
+   * allocated.
+   *
+   * @param bytes The size of the allocation
+   * @param alignment Alignment of the allocation
+   * @return void* Pointer to the newly allocated memory
+   *---------------------------------------------------------------------------**/
+  void* do_allocate(std::size_t bytes, std::size_t alignment =
+                                           alignof(std::max_align_t)) override {
+#if __cplusplus >= 201703L
+    return ::operator new(bytes, std::align_val_t(alignment));
+#else
+    // don't allocate anything if the user requested zero bytes
+    if (0 == bytes) {
+      return nullptr;
+    }
+    // allocate memory for bytes, plus potential alignment correction,
+    // plus store of the correction offset
+    void *p = ::operator new(bytes + alignment + sizeof(std::size_t));
+    std::size_t ptr_int = reinterpret_cast<std::size_t>(p);
+    // calculate the offset, i.e. how many bytes of correction were necessary
+    // to get an aligned pointer
+    std::size_t offset =
+        (ptr_int % alignment) ? (alignment - ptr_int % alignment) : 0;
+    // calculate the return pointer
+    char *ptr = static_cast<char *>(p) + offset;
+    // store the offset right after the actually returned value
+    std::size_t *offset_store = reinterpret_cast<std::size_t *>(ptr + bytes);
+    *offset_store = offset;
+    return static_cast<void *>(ptr);
+#endif
+  }
+
+  /**---------------------------------------------------------------------------*
+   * @brief Deallocate memory pointed to by `p`.
+   *
+   * `p` must have been returned by a prior call to `allocate(bytes,alignment)`
+   * on a `host_memory_resource` that compares equal to `*this`, and the storage
+   * it points to must not yet have been deallocated, otherwise behavior is
+   * undefined.
+   *
+   * @throws Nothing.
+   *
+   * @param p Pointer to be deallocated
+   * @param bytes The size in bytes of the allocation. This must be equal to the
+   * value of `bytes` that was passed to the `allocate` call that returned `p`.
+   * @param alignment Alignment of the allocation. This must be equal to the
+   *value of `alignment` that was passed to the `allocate` call that returned
+   *`p`.
+   * @param stream Stream on which to perform deallocation
+   *---------------------------------------------------------------------------**/
+  void do_deallocate(
+      void* p, std::size_t bytes,
+      std::size_t alignment = alignof(std::max_align_t)) override {
+#if __cplusplus >= 201703L
+    ::operator delete(p, bytes, std::align_val_t(alignment));
+#else
+    (void)alignment;
+    if(nullptr == p){
+        return;
+    }
+    char *ptr = static_cast<char *>(p);
+    // calculate where the offset is stored
+    std::size_t *offset = reinterpret_cast<std::size_t *>(ptr + bytes);
```

Comment:
Trevor's suggestion applies here too:
```suggestion
    auto const offset = *reinterpret_cast<std::size_t *>(ptr + bytes);
```